### ISSUE TITLE
Transaction is logged even when no change is involved

### ISSUE THREAD

**Issue by ambuznego** [bitcrowd/carbonite#60]
When a mutation to update an entity which has no actual changes is attempted, the transaction is still recorded but it has no change record associated to it. I would expect no transaction to be inserted since `changes` in the `Ecto.Changeset` is `%{}`.

```
[debug] QUERY OK db=0.1ms idle=289.6ms
begin []
[debug] QUERY OK db=3.9ms
INSERT INTO "carbonite_default"."transactions" ("meta","inserted_at") VALUES ($1,$2) ON CONFLICT ("id") DO UPDATE SET "id" = EXCLUDED."id" RETURNING "inserted_at","meta","xact_id","id" [%{type: "user_updated"}, ~U[2022-10-31 16:36:13.186498Z]]
[debug] QUERY OK db=1.1ms
commit []
```

I figured at least `updated_at` for the entity would be logged in the change for the transaction, but that doesn't even seem to be what triggers the transaction. I tried excluding `updated_at` through the trigger configuration, and that does not stop the transaction from being created.

What is the best way to address this?

**Comment by maltoe** [comment]
To add to what @andreasknoepfle said: The behaviour is "intended", at least in so far in our own use-cases we've always considered it acceptable to have these "empty" transactions. In the end the transaction did occur nevertheless, right? We also use it to store metadata like the id of the user who triggered the transaction.

**Comment by maltoe** [comment]
@andreasknoepfle I looked around the code again and making the FK on changes deferrable is not an option because of 

https://github.com/bitcrowd/carbonite/blob/f85db00c17cb092d61307ecb3814bc391c397e69/lib/carbonite/migrations/v5.ex#L112-L119

... in the trigger. Recap why this code exists:

- At some point we discovered that `pg_current_xact_id` as PK on `transactions` is not enough as it is reset for new Postgres installations (= it isn't backed up). We gave `transactions` a normal auto-incrementing `id` column.
- The `id` column by itself is unique (PK), but we still wanted to ensure that changes associated to a transaction are inserted in the same transaction.
- Transactions has a PK index spanning the `xact_id` column, though uniqueness lies only on the `id`. I'm not 100% sure why we didn't make it a composite PK. Changing this could be a way of getting rid of this `SELECT` statement in the trigger.

As long as it is present in the trigger though, deferring the FK doesn't do anything.

---

@ambuznego As said before, we consider these empty transactions "intended behaviour". If you insist of removing them, I'd suggest you delete them at the end of the transaction with something like this:

```SQL
DELETE FROM carbonite_default.transactions WHERE id = (
  SELECT id FROM carbonite_default.transactions t 
  LEFT JOIN carbonite_default.changes c ON t.id = c.transaction_id 
  WHERE t.xact_id = pg_current_xact_id() AND c.id IS NULL
);
```
